gmlc/networking/SocketFactory.h

```diff
@@ -0,0 +1,162 @@
+            if (!verify_file.empty()) {
+                ssl_context.load_verify_file(verify_file);
+            }
+            if (!verify_path.empty()) {
+                ssl_context.add_verify_path(verify_path);
+            }
+            if (!certificate_chain_file.empty()) {
+                ssl_context.use_certificate_chain_file(certificate_chain_file);
+            }
+            if (!certificate_file.empty()) {
+                ssl_context.use_certificate_file(
+                    certificate_file, asio::ssl::context::pem);
+            }
+            if (!private_key_file.empty()) {
+                ssl_context.use_private_key_file(
+                    private_key_file, asio::ssl::context::pem);
+            }
+            if (!rsa_private_key_file.empty()) {
+                ssl_context.use_rsa_private_key_file(
+                    rsa_private_key_file, asio::ssl::context::pem);
+            }
+            if (!tmp_dh_file.empty()) {
+                ssl_context.use_tmp_dh_file(tmp_dh_file);
+            }
+
+            return std::make_shared<
+                AsioSocket<asio::ssl::stream<asio::ip::tcp::socket>>>(
+                io_context, ssl_context);
+#else
+            throw std::runtime_error(
+                "gmlc::networking library not compiled with encryption support");
+#endif
+        }
+    }
+
+    /** set if the socket created should act as a server for handshakes when
+     * establishing a connection
+     *
+     * @param b true if the created socket should act as the server in
+     * handshake, otherwise false
+     */
+    void set_handshake_server(bool b) { handshake_server = b; }
+
+    /** set whether the sockets created by this factory should be encrypted or
+     * not
+     *
+     * @param b true if the sockets should be encrypted, otherwise false
+     */
+    void set_encrypted(bool b) { encrypted = b; }
+
+    /** load settings into the SocketFactory from a JSON config file
+     *
+     * @param file the JSON file to load settings from
+     */
+    void load_json_config_file(std::string file);
+
+    /** load settings into the SocketFactory from a JSON string
+     *
+     * @param conf_str the JSON string to parse settings from
+     */
+    void parse_json_config(std::string conf_str);
+
+  private:
+    // other options that could be created:
+    // - ssl protocol/mode to use (tls, sslv3, etc); though tls should probably
+    // always be used
+    // - verify depth for how far down a certificate chain to go
+    // - verify mode for one sided verification, or looser requirements
+    // - custom verify and password callback functions
+    bool encrypted{false};
+    bool handshake_server{false};  // whether or not the socket should act like
+                                   // a server for SSL handshake purposes
+    bool use_default_verify_paths{
+        false};  // use default directories for finding CA files
+    std::string verify_file;  // CA file used in verification for SSL
```

Comment:
Should most of these variables only be enabled if encryption is compiled?

Reply:
If the networking library is built without encryption, changing the values of most of the variables won't do anything. The variables and json parsing for most of the encryption options could be wrapped in an #ifdef, if you think that would cleaner?

Setting the encryption option to on will result in an error when a socket is created, to prevent users from inadvertently using an unencrypted connection.

Reply:
Mostly thinking about size of the resulting object, but My guess is it will make very little difference regardless and might also produce some warnings about unused variables.

Reply:
Warnings don't seem to be a problem. When compiling without encryption I changed it to exclude most of the SSL encryption specific variables and related config parsing code, other than the `encrypted` option so an error can be thrown, and `handshake_server` in case there is a future protocol (websockets?) or backend needs the extra step to establish a connection.